File: Editorial/1601-1700/1602_Find_Nearest_Right_Node_in_Binary_Tree/CPP/4_Recursive_DFS_Preorder_Traversal.cpp
class Solution {
private:
    int u_depth;
    TreeNode* next_node;
    TreeNode* target_node;

public:
    TreeNode* findNearestRightNode(TreeNode* root, TreeNode* u) {
        this->u_depth = -1;
        this->target_node = u;
        this->next_node = nullptr;
        dfs(root, 0);
        return this->next_node;
    }

    void dfs(TreeNode* curr_node, int depth) {
        // the depth to look for next node is identified
        if (curr_node == this->target_node) {
            this->u_depth = depth;
            return;
        }

        // we're on the level to look for the next node
        if (depth == this->u_depth) {
            if (this->next_node == nullptr) {
                this->next_node = curr_node;
            }
            return;
        }

        // continue to traverse the tree
        if (curr_node->left != nullptr) {
            dfs(curr_node->left, depth + 1);
        }
        if (curr_node->right != nullptr) {
            dfs(curr_node->right, depth + 1);
        }
    }
};

File: Editorial/1601-1700/1602_Find_Nearest_Right_Node_in_Binary_Tree/Python3/4_Recursive_DFS_Preorder_Traversal.py
class Solution:
    def findNearestRightNode(self, root: TreeNode, u: TreeNode) -> TreeNode:
        def dfs(current_node, depth):
            nonlocal u_depth, next_node
            # the depth to look for next node is identified
            if current_node == u:
                u_depth = depth
                return
            # we're on the level to look for the next node
            if depth == u_depth:
                # if this next node is not identified yet
                if next_node is None:
                    next_node = current_node
                return
            # continue to traverse the tree
            if current_node.left:
                dfs(current_node.left, depth + 1)
            if current_node.right:
                dfs(current_node.right, depth + 1)

        u_depth, next_node = -1, None
        dfs(root, 0)
        return next_node
